pkgs/build-support/rust/build-rust-package/default.nix

```
@@ -15,145 +15,149 @@
-  # the "${}" is needed to transform the path into a /nix/store path before baseNameOf
-  shortTarget = if targetIsJSON then
-      (lib.removeSuffix ".json" (builtins.baseNameOf "${target}"))
-    else target;
-
-  sysroot = callPackage ./sysroot { } {
-    inherit target shortTarget;
-    RUSTFLAGS = args.RUSTFLAGS or "";
-    originalCargoToml = src + /Cargo.toml; # profile info is later extracted
-  };
-
-in
-
-# Tests don't currently work for `no_std`, and all custom sysroots are currently built without `std`.
-# See https://os.phil-opp.com/testing/ for more information.
-assert useSysroot -> !(args.doCheck or true);
-
-stdenv.mkDerivation ((removeAttrs args [ "depsExtraArgs" "cargoUpdateHook" "cargoLock" ]) // lib.optionalAttrs useSysroot {
-  RUSTFLAGS = "--sysroot ${sysroot} " + (args.RUSTFLAGS or "");
-} // {
-  inherit buildAndTestSubdir cargoDeps;
 
-  cargoBuildType = buildType;
-
-  cargoCheckType = checkType;
-
-  cargoBuildNoDefaultFeatures = buildNoDefaultFeatures;
-
-  cargoCheckNoDefaultFeatures = checkNoDefaultFeatures;
-
-  cargoBuildFeatures = buildFeatures;
-
-  cargoCheckFeatures = checkFeatures;
-
-  patchRegistryDeps = ./patch-registry-deps;
-
-  nativeBuildInputs = nativeBuildInputs ++ [
-    cacert
-    git
-    cargoBuildHook
-    cargoCheckHook
-    cargoInstallHook
-    cargoSetupHook
-    rustc
-  ];
-
-  buildInputs = buildInputs
+# depsExtraArgs, cargoLock have to be removed otherwise 'cannot coerce a set to a string'
+# with structuredAttrs this won't be needed
+# also cargoUpdateHook to avoid changing hashes
+(stdenv.mkDerivation (removeAttrs args [ "depsExtraArgs" "cargoLock" "cargoUpdateHook" ])).overrideAttrs (finalAttrs: previousAttrs:
+  let
+    # If we have a cargoSha256 fixed-output derivation, validate it at build time
+    # against the src fixed-output derivation to check consistency.
+    # NOTE: this seems unused
+    #validateCargoDeps = finalAttrs ? cargoHash || finalAttrs ? cargoSha256;
+
+    target = rust.toRustTargetSpec stdenv.hostPlatform;
+    targetIsJSON = lib.hasSuffix ".json" target;
+    useSysroot = targetIsJSON && !__internal_dontAddSysroot;
+
+    # see https://github.com/rust-lang/cargo/blob/964a16a28e234a3d397b2a7031d4ab4a428b1391/src/cargo/core/compiler/compile_kind.rs#L151-L168
+    # the "${}" is needed to transform the path into a /nix/store path before baseNameOf
+    shortTarget =
+      if targetIsJSON then
+        (lib.removeSuffix ".json" (builtins.baseNameOf "${target}"))
+      else target;
+    sysroot = callPackage ./sysroot { } {
+      inherit target shortTarget;
+      RUSTFLAGS = finalAttrs.RUSTFLAGS or "";
+      originalCargoToml = finalAttrs.src + /Cargo.toml; # profile info is later extracted
+    };
+
+    cargoDeps =
+      if cargoVendorDir != null then null
+      else if cargoLock != null then importCargoLock cargoLock
+      else
+        fetchCargoTarball ({
+          src = finalAttrs.src or "";
```

```suggestion
          src = finalAttrs.src or null;
```
probably not meaningful anyway, but keeping the default unchanged seems nicer